modules/dnn/src/layers/detection_output_layer.cpp

```diff
@@ -211,11 +212,154 @@ class DetectionOutputLayerImpl : public DetectionOutputLayer
+                {
+                    int startIdx = p * num_loc_classes * 4;
+                    util::NormalizedBBox& bbox = decode_bboxes[label][p];
+                    bbox.xmin = decode_data[startIdx + c * 4];
+                    bbox.ymin = decode_data[startIdx + c * 4 + 1];
+                    bbox.xmax = decode_data[startIdx + c * 4 + 2];
+                    bbox.ymax = decode_data[startIdx + c * 4 + 3];
+                }
+            }
+        }
+        return true;
+    }
+
+    void ocl_GetConfidenceScores(const UMat& inp1, const int num,
+                                 const int numPredsPerClass, const int numClasses,
+                                 std::vector<Mat>& confPreds)
+    {
+        int shape[] = { numClasses, numPredsPerClass };
+        for (int i = 0; i < num; i++)
+            confPreds.push_back(Mat(2, shape, CV_32F));
+
+        UMat umat = inp1.reshape(1, num * numPredsPerClass);
+        for (int i = 0; i < num; ++i)
+        {
+            Range ranges[] = { Range(i * numPredsPerClass, (i + 1) * numPredsPerClass), Range::all() };
+            umat(ranges).t().copyTo(confPreds[i]);
+        }
+    }
+
+    bool forward_ocl(InputArrayOfArrays inps, OutputArrayOfArrays outs, OutputArrayOfArrays internals)
+    {
+        std::vector<UMat> inputs;
+        std::vector<UMat> outputs;
+
+        inps.getUMatVector(inputs);
+        outs.getUMatVector(outputs);
+
+        std::vector<LabelBBox> allDecodedBBoxes;
+        std::vector<Mat> allConfidenceScores;
+
+        int num = inputs[0].size[0];
+
+        // extract predictions from input layers
+        {
+            int numPriors = inputs[2].size[2] / 4;
+
+            // Retrieve all confidences
+            ocl_GetConfidenceScores(inputs[1], num, numPriors, _numClasses, allConfidenceScores);
+
+            // Decode all loc predictions to bboxes
+            bool ret = ocl_DecodeBBoxesAll(inputs[0], inputs[2], num, numPriors,
+                                           _shareLocation, _numLocClasses, _backgroundLabelId,
+                                           _codeType, _varianceEncodedInTarget, false,
+                                           allDecodedBBoxes);
+            if (!ret)
+                return false;
+        }
+
+        size_t numKept = 0;
+        std::vector<std::map<int, std::vector<int> > > allIndices;
+        for (int i = 0; i < num; ++i)
+        {
+            numKept += processDetections_(allDecodedBBoxes[i], allConfidenceScores[i], allIndices);
+        }
+
+        if (numKept == 0)
+        {
+            // Set confidences to zeros.
+            Range ranges[] = {Range::all(), Range::all(), Range::all(), Range(2, 3)};
+            outputs[0](ranges).setTo(0);
+            return true;
+        }
+        int outputShape[] = {1, 1, (int)numKept, 7};
+        UMat umat = UMat(4, outputShape, CV_32F);
+        float* outputsData = umat.getMat(ACCESS_READ).ptr<float>();
```

Comment:
This line is not correct:
- umat.getMat() maps OpenCL buffer into `cv::Mat`
- get ptr from `cv::Mat`
- destroy temporary `cv::Mat` and unmap all data back to OpenCL buffer
- pointer is invalid and can't be used.

ACCESS_READ => WRITE

Reply:
You need to create code scope for mapped map, to sync data back before populating of UMat's outputs.
```
{
    Mat mat = umat.getMat(ACCESS_WRITE);
    ... work with mat (umat is blocked)
} // sync data mat->umat, unlock umat
... continue to work with umat ...
```

Reply:
Thanks for the review,  have updated the code with this comments.